src/Elasticsearch/Metadata/Resource/Factory/ElasticsearchOperationProviderResourceMetadataCollectionFactory.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Elasticsearch\Metadata\Resource\Factory;
+
+use ApiPlatform\Elasticsearch\State\CollectionProvider;
+use ApiPlatform\Elasticsearch\State\ItemProvider;
+use ApiPlatform\Elasticsearch\State\OptionsInterface;
+use ApiPlatform\Metadata\CollectionOperationInterface;
+use ApiPlatform\Metadata\GraphQl\Operation as GraphQlOperation;
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\Metadata\Operations;
+use ApiPlatform\Metadata\Resource\Factory\ResourceMetadataCollectionFactoryInterface;
+use ApiPlatform\Metadata\Resource\ResourceMetadataCollection;
+
+final class ElasticsearchOperationProviderResourceMetadataCollectionFactory implements ResourceMetadataCollectionFactoryInterface
+{
+    public function __construct(
+        private readonly ResourceMetadataCollectionFactoryInterface $decorated,
+    ) {
+    }
+
+    public function create(string $resourceClass): ResourceMetadataCollection
+    {
+        $resourceMetadataCollection = $this->decorated->create($resourceClass);
+
+        foreach ($resourceMetadataCollection as $i => $resourceMetadata) {
+            if ($operations = $resourceMetadata->getOperations()) {
+                $resourceMetadataCollection[$i] = $resourceMetadata->withOperations($this->configureOperations($operations));
+            }
+            if ($graphQlOperations = $resourceMetadata->getGraphQlOperations()) {
+                $resourceMetadataCollection[$i] = $resourceMetadata->withGraphQlOperations($this->configureOperations($graphQlOperations));
+            }
+        }
+
+        return $resourceMetadataCollection;
+    }
+
+    /**
+     * @template T of Operations|array<GraphQlOperation>
+     *
+     * @phpstan-param T $operations
+     *
+     * @phpstan-return T
+     */
+    private function configureOperations(Operations|array $operations): Operations|array
+    {
+        foreach ($operations as $operationName => $operation) {
+            $configuredOperation = $this->configureOperation($operation);
+            if ($configuredOperation === $operation) {
+                continue;
+            }
+            if ($operations instanceof Operations) {
+                $operations->add($operationName, $configuredOperation);
+                continue;
+            }
+
+            $operations[$operationName] = $configuredOperation;
+        }
+
+        return $operations;
+    }
+
+    private function configureOperation(Operation $operation): Operation
+    {
+        if (null !== $operation->getProvider()) {
+            return $operation;
+        }
+        if (!$operation->getStateOptions() instanceof OptionsInterface) {
```

Comment:
I don't understand this, if we set a state options we don't set the provider?

Reply:
In this code, we set an elasticsearch provider only if `stateOptions` is an instance of `ApiPlatform/Elasticsearch/State/OptionsInterface`, isn't that what we want?

Reply:
okay yes and this is what I find weird about this implementation instead of a specific Elasticsearch Operation. What is more explicit for the user:

```
use ApiPlatform\Elasticsearch\State\Options;
use ApiPlatform\Metadata\GetCollection;

#[GetCollection(stateOptions: new Options(index: 'test'))
class MyResource {}
```

or


```
use ApiPlatform/Elasticsearch\Metadata\GetCollection;

#[ApiResource(index: 'test')
class MyResource {}
```